### PR TITLE
0.1.0 implementation

### PR DESCRIPTION
**Pull request by ayrton**
The initial `0.1.0` implementation

### CODE REVIEW

File: demo/components/app/index.js [hunk 1]
```diff
@@ -10,8 +10,16 @@ import Cluster from '../../../lib';
...
-      <Cluster />
-    </div>
+    <Cluster className="cluster" height={300} rowHeight={30}>
+      {renderItems()}
+    </Cluster>
+  );
+}
+
+function renderItems(): Array<ReactElement> {
+  const items = Array.apply(null, Array(10000));
```
Comment (andreasklinger) [inline demo/components/app/index.js]: why?

File: lib/components/cluster.js [hunk 2]
```diff
@@ -3,11 +3,116 @@
...
+export default class Cluster extends React.Component {
+  props: Props;
+  state: State = INITIAL_STATE;
+
+  shouldComponentUpdate(nextProps: Props, nextState: State): boolean {
+    return this.props.children !== nextProps.children || this.state.index !== nextState.index;
+  }
+
+  componentDidMount(): void {
+    this.cluster.addEventListener('scroll', this.setIndex.bind(this));
```
Comment (andreasklinger) [inline lib/components/cluster.js]: please use `this.refs` - i searched way to long for the constructor

File: lib/components/cluster.js [hunk 3]
```diff
@@ -3,11 +3,116 @@
...
+
+    return (
+      <div className={className} ref={(ref) => this.cluster = ref} style={{ height, overflow: 'scroll' }}>
+        {rows}
+      </div>
+    );
+  }
+
+  findIndex(): number {
+    return Math.floor(this.cluster.scrollTop / this.props.rowHeight);
```
Comment (andreasklinger) [inline lib/components/cluster.js]: this calculates which index is the current one right?

why is it using rowheight not something like rowheight*childrencount ?

File: lib/components/cluster.js [hunk 4]
```diff
@@ -3,11 +3,116 @@
...
+  }
+
+  findRange() : Array<number> {
+    const {height, rowHeight} = this.props;
+    const {index} = this.state;
+
+    const start = index - RANGE_OFFSET;
```
Comment (andreasklinger) [inline lib/components/cluster.js]: RANGE_OFFSET = VISIBLE_ROWS (/2) ?

File: lib/components/cluster.js [hunk 5]
```diff
@@ -3,11 +3,116 @@
...
+  }
+
+  render(): ReactElement {
+    const {children, className, height, rowHeight} = this.props;
+
+    const rows = toClusterRows({
+      children,
+      height: rowHeight,
+      range: this.findRange(),
+    });
```
Comment (andreasklinger) [inline lib/components/cluster.js]: this feels overly complex

eg 
- why does range pass findRange (vs having it in the function directly)
- why is this not simply a function that's called (eg renderRows)

File: lib/components/cluster.js [hunk 6]
```diff
@@ -3,11 +3,116 @@
...
+  let index = -1;
+  let result = Array(length);
+
+  while (++index < length) {
+    result[index] = start;
+    start += 1;
+  }
+
+  return result;
+}
```
Comment (andreasklinger) [inline lib/components/cluster.js]: this feels overly complex

eg 
- why result[index] vs array.push
- why not using `_.range` 

but tbh range feels like an overkill

while i would do this the same way in ruby it's too much for here
you are doing this to be able to do `.includes()` below

i would do 

```
const [firstVisibleRow, lastVisibleRow] = this.findRange();
…
(index > firstVisibleRow && index < lastVisibleRow) ? child : null
```

File: lib/components/cluster.js [hunk 7]
```diff
@@ -3,11 +3,116 @@
...
+  componentDidMount(): void {
+    this.cluster.addEventListener('scroll', this.setIndex.bind(this));
+  }
+
+  componentWillUnmount(): void {
+    this.cluster.removeEventListener('scroll', this.setIndex.bind(this));
+  }
+
+  setIndex(): void {
+    this.setState({ index: this.findIndex() });
```
Comment (andreasklinger) [inline lib/components/cluster.js]: maybe `this.setCurrentIndex()` ?

File: lib/components/cluster.js (hunk 2 above)
Comment (ayrton) [inline lib/components/cluster.js]: AFAIK that's no longer [the recommended way and this is](https://facebook.github.io/react/docs/more-about-refs.html).

File: lib/components/cluster.js (hunk 3 above)
Comment (ayrton) [inline lib/components/cluster.js]: That would give you a px offset, while we need an index

File: lib/components/cluster.js (hunk 7 above)
Comment (ayrton) [inline lib/components/cluster.js]: :+1:

**Comment by andreasklinger** [comment]
:+1:

**Pull request merged**